Minor changes

DavidJohnGee opened the pull request:
1.  Description change for the load_config action. Improved for ease of use without guess work.

2. Addition of 'inline_transfer' parameter. This is specifically for the Cisco IOS driver. Does not affect other drivers that does not use the 'inline_transfer' field.

DavidJohnGee committed:
Required for Cisco IOS oddities

Minor change for Cisco IOS driver. If inline_transfer is True, then the Cisco IOS driver (Netmiko) will use the SSH channel to transfer the configuration change. If the inline_transfer is False (unset), then the Cisco IOS driver will attempt to use SCP.

For some router operating systems, attaining a suitable privilege level is fraught with trouble. Inlining is the easiest way to solve the issue in some cases.

For other drivers (non 'ios'), unless the driver has an internal 'inline_transfer' field, it has no affect. By default is not set and is not required.

DavidJohnGee committed:
inline_transfer addition

This input parameter allows users to set config transfers to either be via SSH or SCP.

Although I'm using the 'TRUE' string here, the parameter in the Cisco IOS is a boolean. Therefore if it's passed to the parameter, it evaluates as True, if it is not set, the parameter is evaluated to be False; it becomes true from a side-effect as opposed to direct manipulation of the boolean. 

This is specifically covering the Cisco IOS driver at this time. Adding this parameter might also be useful for other drivers.

DavidJohnGee committed:
Update to config_file parameter desc

Minor update to config_file description field to be more helpful.

DavidJohnGee committed:
Update loadconfig.py

Mended trailing whitespace.

DavidJohnGee committed:
Merge branch 'master' into master

DavidJohnGee committed:
Update loadconfig.py

New line end of file added (d'oh).

DavidJohnGee committed:
Update loadconfig.py

Web editor is a bit janky. Whitespace sorted out.

Mierdin commented on actions/loadconfig.py:
```diff
@@ -5,22 +5,31 @@ class NapalmLoadConfig(NapalmBaseAction):
-        with self.get_driver(**std_kwargs) as device:
+            with self.get_driver(**std_kwargs) as device:
+                # inline_transfer: If set it becomes True, else False
+                device.inline_transfer = inline_transfer
```
I'm not familiar with `inline_transfer` - what is it?

DavidJohnGee replied:
inline_transfer is specific to Netmiko. This allows the transfer of configuration within the SSH channel (i.e. CLI) as opposed to initialising an SCP session. Depending on the configuration of the router, this can fail and cause issues. I want to give users the option (in most cases to avoid forcing users to make changes).

Mierdin commented on actions/loadconfig.py:
```diff
@@ -5,22 +5,31 @@ class NapalmLoadConfig(NapalmBaseAction):
+
+                device.commit_config()
+
+        except Exception, e:
```
This block was removed in #34. It looks like you have the commits from that incorporated already, so maybe something else happened in your editor? Anyways, please ensure this is removed per the work in #34

DavidJohnGee replied:
Yep - will take care of this. 

I'm not quite sure what happened with this. I pulled the latest version to work on and it had the try/except statements in. When looking on the Web UI, they're not there. Will try again and see what's going on.

Mierdin commented on actions/loadconfig.yaml:
```diff
@@ -27,11 +27,18 @@ parameters:
     config_file:
         type: "string"
-        description: "configuration to load to device"
+        description: "Configuration file including full path to load to device (/x/y/z.type)"
```
How about just "Absolute path to configuration file on local system that will be sent to the device"

DavidJohnGee replied:
That works.

Mierdin commented on actions/loadconfig.yaml:
```diff
@@ -27,11 +27,18 @@ parameters:
         description: "Method to load the config (either 'merge' or 'replace', default is 'merge')"
         required: false
         position: 5
+    inline_transfer:
```
I'd prefer this is `boolean` type. What is the expected type of the `device.inline_transfer` property?

DavidJohnGee replied:
Have to admit to being lazy here. I tried boolean and couldn't get it to work at the time. Went for speed over correctness. I will punish myself appropriately.

automationspace committed:
Changed commit to be inline with requests from: https://github.com/StackStorm-Exchange/stackstorm-napalm/pull/35

1) Changed config_file description
2) Changed inline_transfer to boolean
3) Removed loadconfig.py try/except statements to be inline with pull 34
4) Bumped pack version

automationspace committed:
Added config_file description change

DavidJohnGee commented:
Ok. Changes made! Second review?

Mierdin commented on actions/loadconfig.py:
```diff
@@ -16,12 +16,14 @@ def run(self, config_file, method, **std_kwargs):
                 device.load_merge_candidate(filename=config_file)
 
-            device.commit_config()
+                device.commit_config()
```
Accidental indentation?

DavidJohnGee replied:
Damn it. Will amend.

Mierdin commented on actions/loadconfig.py:
```diff
@@ -16,12 +16,14 @@ def run(self, config_file, method, **std_kwargs):
 
         with self.get_driver(**std_kwargs) as device:
+            # inline_transfer: If set it becomes True, else False
+            device.inline_transfer = inline_transfer
```
I reviewed the `inline_transfer` implementation in the NAPALM IOS driver, and this almost seems like it would be better to implement this in the config, and pass in via `optional_args`. Doing this would re-use the built-in logic in the driver for deciding what the default value for this parameter should be - assigning it on the instance like this bypasses that logic in the driver's `__init__` function. It would also prevent us from having to make changes if they bring that functionality to other drivers. See [here](https://github.com/StackStorm-Exchange/stackstorm-napalm/blob/master/actions/lib/action.py#L44) for an example, we're already doing this with port. Thoughts?

DavidJohnGee replied:
Part of me agrees and part of me disagrees. 

Agree: Getting the `__init__()` function to do this from the config (per host entry) would be tidy. It did cross my mind originally but decided against it. 

Disagree: I've had mixed experience with different versions of OS that the ios driver communicates with. If it's an action decision as opposed to config, then if the action fails, we can  try again on error transition to try inline as opposed to SCP. 

Right now I'm leaning 60/40 in favour of action over config. Happy to take your lead on this.

Thoughts?

Mierdin replied:
If that's true, then the inconsistency should *definitely* get fixed at the NAPALM layer. Definitely open an issue with them if you're seeing differences in OS behavior.

Since the other drivers effectively ignore this for now, we can leave it in the action - but if the underlying behavior changes, or if they change how the default gets set, this may have to move.

Mierdin commented:
@DavidJohnGee Also might want to review your git config and/or the email addresses you registered with Github. The last few commits don't seem to track to your user account - wouldn't want you to miss out on credit! :smile:

automationspace committed:
Added config_file description change

automationspace committed:
Merge branch 'master' of https://github.com/DavidJohnGee/stackstorm-napalm

automationspace committed:
Amended accidental indentation.
